### PR TITLE
keys/query should include device display name

### PR DESCRIPTION
**Pull request by richvdh**
Per https://github.com/matrix-org/synapse/pull/978

### CODE REVIEW

File: tests/10apidoc/12device_management.pl [hunk 1]
```diff
@@ -10,6 +10,19 @@ sub matrix_get_device {
    );
 }
 
+sub matrix_set_device_display_name {
```
Comment (leonerd) [inline tests/10apidoc/12device_management.pl]: Hrm. This brace should have been on its own line, but then so should the other functions in the file. Don't worry about it; I'll do a style fixup later.

File: tests/41end-to-end-keys/01-upload-key.pl [hunk 2]
```diff
@@ -53,14 +53,29 @@
...
+
+         my $alice_device_keys = $alice_keys->{ $user->device_id };
+         assert_json_keys( $alice_device_keys, "unsigned" );
+
+         my $unsigned = $alice_device_keys->{unsigned};
+
+         # display_name should be null by default
+         exists $unsigned->{device_display_name} or
+           die "Expected to get a (null) device_display_name";
+         defined $unsigned->{device_display_name} and
```
Comment (leonerd) [inline tests/41end-to-end-keys/01-upload-key.pl]: This tests that the key exists but has an undefined value. Are we OK with adding that kind of distinction? I seem to recall it being quite a rare thing in our API that we make use of `null` values in JSON. Maybe @NegativeMjark has a thought here?

File: tests/41end-to-end-keys/04-query-key-federation.pl [hunk 3]
```diff
@@ -27,7 +29,14 @@
 
          my $alice_keys = $device_keys->{ $user->user_id };
          assert_json_keys( $alice_keys, $user->device_id );
+
+         my $alice_device_keys = $alice_keys->{ $user->device_id };
+
          # TODO: Check that the content matches what we uploaded.
+
+         assert_eq( $alice_device_keys->{"unsigned"}->{"device_display_name"},
```
Comment (leonerd) [inline tests/41end-to-end-keys/04-query-key-federation.pl]: FYI, second arrow here is unnecessary; perl doesn't care after the first.

File: tests/41end-to-end-keys/01-upload-key.pl (hunk 2 above)
Comment (NegativeMjark) [inline tests/41end-to-end-keys/01-upload-key.pl]: I dunno. I'd guess that we should handle this in the same way that user displaynames are handled.

Comment (richvdh) [inline tests/41end-to-end-keys/01-upload-key.pl]: Yeah, fair enough. The instances of that which I've checked in the spec omit the key when no displayname is set.

Comment (richvdh) [inline tests/41end-to-end-keys/01-upload-key.pl]: updated in https://github.com/matrix-org/synapse/pull/981.

**Comment by richvdh** [comment]
(tests are failing because the fix is on a different branch)

**Comment by richvdh** [comment]
matrixbot: please retest

**Comment by richvdh** [comment]
matrixbot: retest this please

even

**Pull request merged**